### PR TITLE
Make parse_ranges_unsigned actually return an unsigned range

### PR DESCRIPTION
**Pull request by Vultraz**
This shouldn't result in any behavior changes, just makes the typing more strict. Also cleans up some of the relevant code using in_ranges. It does raise the question as to whether a lot of these variables should be unsigned instead of int, but that's a different matter altogether...

### CODE REVIEW

File: src/utils/config_filters.cpp [hunk 1]
```diff
@@ -71,7 +71,7 @@ bool utils::config_filters::unsigned_matches_if_present(const config& filter, co
 		return false;
 	}
 
-	return in_ranges<int>(cfg[attribute].to_int(0), utils::parse_ranges_unsigned(filter[attribute].str()));
+	return in_ranges<unsigned>(cfg[attribute].to_unsigned(0), utils::parse_ranges_unsigned(filter[attribute].str()));
```
Comment (Pentarctagon) [inline src/utils/config_filters.cpp:74]: This seems wrong? Attributes can have negative values.

Comment (Vultraz) [reply src/utils/config_filters.cpp:74]: Would that mean the original code was buggy? Because a negative value would always fail regardless.

Comment (Pentarctagon) [reply src/utils/config_filters.cpp:74]: Well, I suppose it doesn't matter either way given the function is unused:
```
grep -Iirn 'unsigned_matches_if_present'
utils/config_filters.cpp:65:bool utils::config_filters::unsigned_matches_if_present(const config& filter, const config& cfg, const std::string& attribute)
utils/config_filters.hpp:52:bool unsigned_matches_if_present(const config& filter, const config& cfg, const std::string& attribute);
tests/test_config_filters.cpp:31:       // but for the unsigned_matches_if_present function only one of them is
tests/test_config_filters.cpp:36:       BOOST_ASSERT(!unsigned_matches_if_present(minus_3, minus_3, "x"));
tests/test_config_filters.cpp:37:       BOOST_ASSERT(unsigned_matches_if_present(plus_3, plus_3, "x"));
tests/test_config_filters.cpp:39:       BOOST_ASSERT(!unsigned_matches_if_present(minus_3, plus_3, "x"));
tests/test_config_filters.cpp:40:       BOOST_ASSERT(!unsigned_matches_if_present(plus_3, minus_3, "x"));
```
I think it's worth considering what would happen in other cases though if the value did actually underflow to the max value of `unsigned`.

Comment (Vultraz) [reply src/utils/config_filters.cpp:74]: That does make me realize one small hole in these changes (and I guess the original implementation). `infinity` is represented as the max value of `int`... which is half the max value of `unsigned`. 🤔 One would expect parsing an unsigned range to infinity would include that extra space.

**Comment by stevecotton** [comment]
This is touching abilities.cpp and attack_type.cpp, which are places of active work that'll need some parts to change in 1.18 (the abilities infinite recursion bugs). Some of the lines about '"count"' attributes are currently having tests written to work out whether they're broken, which is happening in #9074 and #9079.

Please hold off on unnecessary changes to the same area until we've worked out what needs fixing in 1.18.

**Comment by Vultraz** [comment]
I’ll leave this to you to merge when appropriate, then.

**Comment by CelticMinstrel** [comment]
> This shouldn't result in any behavior changes

This is a naïve statement, as changing signed to unsigned could _absolutely_ produce behaviour changes if there were cases where the numbers in question were incorrectly negative… and given that at least some of these surely come from user input, the chance of those cases existing is probably 100%.

**Comment by Vultraz** [comment]
The code already checks whether the lower bound is negative.

**Comment by CelticMinstrel** [comment]
Well, as long as the negative check is done on signed numbers (or before parsing it), that sounds like it should probably be fine…

**Comment by Vultraz** [comment]
Yes, that’s what happens, and I didn’t change that.

**Comment by CelticMinstrel** [comment]
> That does make me realize one small hole in these changes (and I guess the original implementation). `infinity` is represented as the max/min value of `int`... which is half the max value of `unsigned`. 🤔 One would expect parsing an unsigned range to infinity would include that extra space.

This at least seems worth changing though.

**Comment by Vultraz** [comment]
Should be ready to merge now.

**Comment by CelticMinstrel** [comment]
This is kinda weird. A signed range can have bounds in the range [-32768,32768), for example. An unsigned range can have bounds in the range [0, 32768) &#x222a; {65535}. There's an entire dead zone where you can't possibly put the bounds.

In practice, I doubt it really matters, as the reality of the situation is probably that these are 32-bit or even 64-bit integers, so while the principle is still true, the actual numbers are so high that you'll basically never hit them. But it's still weird…

**Comment by Pentarctagon** [comment]
@stevecotton any reason to hold off merging this? The abilities changes seem to be at a bit of a standstill at the moment anyway.

**Comment by stevecotton** [comment]
Although there's some good cleanup in here (using `in_ranges` instead of a `for` loop), I think the main change shouldn't be done.

Having a single C++ type to represent ranges from WML seems cleaner to me; sometimes we know that negative values indicate a bug in the WML, so it makes sense to have the parsing function notice that, but I see no reason why that means that the internal representation should change. The suggestion of changing what `infinity` parses to based on whether it's in a range that accepts negative values or not seems to be unnecessary complexity, but it also seems to be almost necessary if the internal representation is changed; so my answer would be to leave the internal representation as-is.

Admittedly, that does leave a function with "unsigned" in its name which returns signed values. I'm sure that at one stage it was going to be called "parse_ranges_positive", but it was decided that "unsigned" was clearer.

**Comment by Vultraz** [comment]
I can split the cleanup off separately.

**Pull request closed without merging**